src/CompiledMatcher.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/*
+ * This file is part of composer/semver.
+ *
+ * (c) Composer <https://github.com/composer>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace Composer\Semver;
+
+use Composer\Semver\Constraint\CompilableConstraintInterface;
+use Composer\Semver\Constraint\Constraint;
+use Composer\Semver\Constraint\ConstraintInterface;
+use Composer\Semver\Constraint\NotCompilableConstraintException;
+
+/**
+ * Helper class to evaluate constraint by compiling and reusing the code to evaluate
+ */
+class CompiledMatcher
+{
+    private static $compiledCheckerCache = array();
+    private static $enabled = null;
+
+    private static $transOpInt = array(
+        Constraint::OP_EQ => '==',
+        Constraint::OP_LT => '<',
+        Constraint::OP_LE => '<=',
+        Constraint::OP_GT => '>',
+        Constraint::OP_GE => '>=',
+        Constraint::OP_NE => '!=',
+    );
+
+    /**
+     * Evaluates the expression: $constraint match $operator $version
+     *
+     * @param ConstraintInterface $constraint
+     * @param int                 $operator
+     * @param string              $version
+     *
+     * @return mixed
+     */
+    public static function match(ConstraintInterface $constraint, $operator, $version)
+    {
+        if (self::$enabled === null) {
+            self::$enabled = !in_array('eval', explode(',', ini_get('disable_functions')));
+        }
+        if (!self::$enabled || !$constraint instanceof CompilableConstraintInterface) {
+            return $constraint->matches(new Constraint(self::$transOpInt[$operator], $version));
+        }
+
+        $cacheKey = $operator.$constraint;
+        if (!isset(self::$compiledCheckerCache[$cacheKey])) {
+            try {
+                $code = $constraint->compile($operator);
+                self::$compiledCheckerCache[$cacheKey] = $function = eval('return function($v, $b){return '.$code.';};');
```

Comment:
Instead of using eval? I mean you can call eval inside the compile() method instead of here, but you're going to need to actually turn the php code string into something executable somewhere? I think it's nice to have the compile method around to access the code directly and potentially write it out elsewhere for other purposes, so this is good as-is?

Reply:
Maybe I am missing something... the only constraint which I can see within this PR which requires the `eval()` is the multi-constraint.. all others dont seem to generate dynamic code from strings, right?

Reply:
Scratch that: the problem is the recursive nature of MultiConstraint->compile()

Reply:
hmm.. That's a goof idea, but I'm not sure.
From one hand, It'll remove this eval. but from another hand, closures returned by Constraint are not mergeable in MultiConstraint. Which means:
- it will generate a lot of closures: one per constraint + one per multi (even if the same constraint it used by several Multi)
- it would reduce performances, by calling several closures instead of 1 that evaluate the complete rule
- this close the door to future enhancement using the file_system instead of eval.

WDYT?

Reply:
Yeah I realized that it will not work in the full picture :).

At best we merge as is and incremental improve with another PR

Reply:
I think the idea of @staabm was:
```php
function compile() {
  if (self::OP_NE === $this->operator) {
    if ($isBranch) {
      if ($otherOperator === self::OP_EQ) {
        return fuction() {return false;};
      }
      return fuction() {return true;};
    }
    if ($otherOperator === self::OP_EQ) {
      $version = $this->version;
      return function($b, $v) use ($version) {
        return !$b && version_compare($v, $version, '!=');
    }
  }
  return fuction() {return true;};
  ...
}

```

used:
```php
$fn = $constraint->compile();

$fn($b, $v);
```

Reply:
Exactly

Reply:
@jderusse I'd say, benchmark it? We're definitely working on something that will do more multi/multi comparisons btw https://github.com/composer/composer/pull/8850

Reply:
Benchmark available [here](https://blackfire.io/profiles/compare/804e3ae2-6c64-4d35-bc09-29b6c0414f38/graph) 
+4% CPU and +1% memory. 

it's not as good as current Impl. but It's not that bad.. 
It remove the "eval" call. But the drawback is we can't persist the compilation result anymore

WDYT 
ping @Toflar @Seldaek

Reply:
Hmmm...Persisting would have the advantage that it could „learn“ with consecutive calls. If we were to keep a list of precompiled constraints in a global Composer cache for example. Evaling from a stored file can be dangerous though so it would require to be optional maybe? The Symfony ExpressionLanguage can compile as well as evaluate. Maybe semver should provide both and Composer/other libs use what‘s best? Not sure if worth the effort though?

Reply:
> it's not as good as current Impl. but It's not that bad..

Thx for checking. I would leave it as is for now

Reply:
@jderusse could you post a profile of the PR as is?

Reply:
profile: https://blackfire.io/profiles/8379ae17-f4c4-4b59-83d7-b7c13bc9f115/graph
comparison with original: https://blackfire.io/profiles/compare/cfa5e43e-0510-4b24-bd44-722dc0c3c7ce/graph